packages/go_router/lib/src/redirection.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'configuration.dart';
+import 'logging.dart';
+import 'match.dart';
+import 'matching.dart';
+
+/// A GoRouter redirector function.
+// TODO(johnpryan): make redirector async (#105808)
+typedef RouteRedirector = RouteMatchList Function(RouteMatchList matches,
+    RouteConfiguration configuration, RouteMatcher matcher,
+    {Object? extra});
+
+/// Processes redirects by returning a new [RouteMatchList] representing the new
+/// location.
+RouteMatchList redirect(RouteMatchList prevMatchList,
+    RouteConfiguration configuration, RouteMatcher matcher,
+    {Object? extra}) {
+  RouteMatchList matches;
+
+  // Store each redirect to detect loops
+  final List<RouteMatchList> redirects = <RouteMatchList>[prevMatchList];
+
+  // Adds the redirect to the list of redirects if it is valid.
+  bool redirected(RouteMatchList newRedirect) {
```

Comment:
IMO this function is harder to read because `redirect` is nested inside. What do you think of unnesting `redirect`, renaming it to something like `_add_redirect`, and adding parameters as needed instead of capturing values from the parent scope? Something like `_add_redirect(redirects, newMatch, prevMatchList.location);`

Reply:
Sounds good to me